handler/src/main/java/io/netty/handler/ssl/SniHandler.java

```diff
@@ -321,6 +321,19 @@ private void replaceHandler(ChannelHandlerContext ctx, Selection selection) {
         }
     }
 
+    /**
+     * Users may override this method to implement custom behavior. The default
+     * implementation will simply replace {@code this} {@link SniHandler} instance
+     * with the given {@link SslHandler} using the provided {@link SSLEngine}.
+     *
+     * BEWARE: The {@code hostname} argument can be {@code null}.
+     */
+    protected void replaceHandler(ChannelHandlerContext ctx, String hostname,
+            SslContext context, SSLEngine engine) throws Exception {
+
+        ctx.pipeline().replace(this, SslHandler.class.getName(), SslContext.newHandler(engine));
```

Comment:
The release responsibility is not clearly transferred to this method. For example if an exception is thrown we release the `engine` out side the scope of this method, but otherwise we assume the release responsibility lies within this method. Also by passing in the `SslContext` a user may be confused and generate another `SSLEngine` and insert this into the pipeline (and thus create a leak). To clarify this I wonder if we should just make this method  `protected void onSelection(String selectedHostName, SslContext selectedSslContext)` do the SSLEngine creation/replacement outside the scope of this method, and indicate in the comments an SSLEngine will be inserted outside the scope of this method by SniHandler. You could insert another handler in the pipeline to get the channel inactive event if necessary ... wdty?